### DISCUSSION TITLE
Order deeply embedded table content

### DISCUSSION THREAD

**Discussion by jhlabs** [supabase/supabase#3012] [Questions]
Hi and thanks for helping me solve this:

I have modeled a graph structure of blocks where each block can have several children and parents. I have a `blocks` table that stores each block with metadata and a `block_relationships` table that stores a `parentId`, `childId`, and a `position` - an alphanumeric index that stores the position of the child in relation to a parent.

I now want to retrieve a tree of blocks starting from a specified block with all children sorted by their `position`.

This is the query I currently have for retrieving the tree for 2 levels (I am using the Supabase Dart Client):

```dart
final result = await client
      .from('blocks')
      .select('''
        *,
        children:block_relations!parentId(child:
        childId(
          *,
          position:block_relations!childId(position),
          children:block_relations!parentId(child:
            childId(
                *,
                position:block_relations!childId(position)
              )
            )
          )
        )
      ''')
      .eq("attributes->>title", title)
      .eq('creatorId', profileId)
      .order('position', foreignTable: 'block_relations', ascending: true)
      .execute();
```

This sorts all children on the first level correctly, but children on the second level remain unordered. I was also considering implementing this structure as a recursive view similar to [this example](https://github.com/lawrencecchen/threaded-comments), but could not make it work for my use case.

Thanks for your help!

**Comment by steve-chavez** [comment]
Can you try using the aliases for ordering the embedded resources? Like:

```dart
final result = await client
      .from('blocks')
      .select('''
        *,
        children:block_relations!parentId(child:
        childId(
          *,
          position:block_relations!childId(position),
          children:block_relations!parentId(child:
            childId(
                *,
                position:block_relations!childId(position)
              )
            )
          )
        )
      ''')
      .eq("attributes->>title", title)
      .eq('creatorId', profileId)
      .order('position', foreignTable: 'children', ascending: true)
      .order('position', foreignTable: 'children.child', ascending: true)
      .execute();
```

> This sorts all children on the first level correctly, but children on the second level remain unordered. I was also considering implementing this structure as a recursive view similar to [this example](https://github.com/lawrencecchen/threaded-comments), but could not make it work for my use case.

That being said, a recursive view looks more fit for this use case.

**Reply by jhlabs** [reply]
@steve-chavez thanks for your quick reply. Using the alias works as well for the first level of nesting, but it does not have any influence on the second level 🤷‍♂️.

I found a way to query for the right data using a recursive table expression, but I have no idea how to convert that to a view and the right query. This is what I wrote to retrieve the items with a hardcoded starting block in the Query Editor in Supabase:

```sql
WITH RECURSIVE block_tree("parentId", "childId", position) AS (
    SELECT "parentId", "childId", position
    FROM block_relations
    JOIN blocks ON blocks.id = block_relations."parentId"
    WHERE attributes->> 'title' = '2021827' AND "creatorId" = '2d809ba6-7740-423b-9825-ff111ec27bd4'
  UNION
    SELECT block_relations."parentId", block_relations."childId", block_relations.position
    FROM block_relations
    JOIN block_tree ON block_tree."childId" = block_relations."parentId"
)

SELECT blocks.id, blocks.type, blocks.attributes, blocks."createdAt", blocks."updatedAt", blocks."creatorId", block_tree.position, block_tree."parentId"
FROM block_tree
JOIN blocks ON blocks.id = block_tree."childId"
UNION
SELECT blocks.id, blocks.type, blocks.attributes, blocks."createdAt", blocks."updatedAt", blocks."creatorId", '0' as position, 'c798bd72-0c52-4656-9067-9dc3080dfcb9' as "parentId"
FROM blocks
WHERE attributes->> 'title' = '2021827'

```

**Reply by steve-chavez** [reply]
>  Using the alias works as well for the first level of nesting, but it does not have any influence on the second level man_shrugging.

@soedirgo Do you think there could be an issue with using nested tables on `foreignTable` in `postgrest-js`? PostgREST does support this https://github.com/PostgREST/postgrest-docs/issues/432.

**Reply by soedirgo** [reply]
Not that I'm aware of, though I think this is using the Dart client? Probably not a problem with the client library either, trying:
```sql
create table table1 (
  id int4 generated by default as identity primary key
);

insert into table1(id) values (1), (2);

create table table2 (
  id int4 generated by default as identity primary key,
  table1_id int4 references table1(id),
  other_table1_id int4 references table1(id)
);

insert into table2(table1_id, other_table1_id) values (1, 1), (1, 2);
```
```js
supabase
  .from('table1')
  .select(`
    table2!table1_id(
      other_table1_id(*)
    )
  `)
```
Gives:
```json
[
    {
      "table2": [
        {
          "other_table1_id": {
            "id": 1
          }
        },
        {
          "other_table1_id": {
            "id": 2
          }
        }
      ]
    },
    {
      "table2": []
    }
]
```
Note that `other_table1_id` is not a list, so `order` doesn't do anything. I think what's needed is something like https://github.com/PostgREST/postgrest/issues/1907?

**Reply by jhlabs** [reply]
@soedirgo yes I am using [supabase_flutter](https://pub.dev/packages/supabase_flutter) which in turn is using the supabase dart client.
